File: baselibrary/src/main/java/com/aslan/baselibrary/http/response/BaseResponseFunction.java
package com.aslan.baselibrary.http.response;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.aslan.baselibrary.http.BaseError;
import com.aslan.baselibrary.http.EventTokenError;
import com.aslan.baselibrary.http.IHttpBean;
import com.aslan.baselibrary.http.NetManager;
import io.reactivex.functions.Function;
import org.greenrobot.eventbus.EventBus;

/**
 * 处理业务异常
 *
 * @author Aslan chenhengfei@yy.com
 * @date 2020/6/12
 */
public abstract class BaseResponseFunction<T, R> implements Function<IHttpBean<T>, R> {

  private Context context;

  public BaseResponseFunction(Context context) {
    this.context = context;
  }

  @Override
  public R apply(IHttpBean<T> respone) throws Exception {
    if (respone.isTokenError()) {
      EventBus.getDefault().post(new EventTokenError(respone.getCode()));
      return error(new BaseError(NetManager.ERROR_TOKEN_ERROR,
          context.getString(com.aslan.baselibrary.R.string.error_net_token_error)));
    }

    if (respone.isSuccessful()) {
      return handleData(respone.getData());
    } else {
      return error(new BaseError(respone.getCode(), respone.getMessage()));
    }
  }

  public abstract R error(@NonNull BaseError ex);

  public abstract R handleData(@Nullable T item);
}

File: baselibrary/src/main/java/com/aslan/baselibrary/http/response/BaseResponseFunction.kt
package com.aslan.baselibrary.http.response

import android.content.Context
import com.aslan.baselibrary.exception.BusinessException
import com.aslan.baselibrary.exception.TokenException
import com.aslan.baselibrary.http.IHttpBean
import io.reactivex.functions.Function

/**
 * 处理业务异常
 *
 * @author Aslan chenhengfei@yy.com
 * @date 2020/6/12
 */
abstract class BaseResponseFunction<T, R>(private val context: Context) :
    Function<IHttpBean<T>, R> {

    @Throws(Exception::class)
    override fun apply(respone: IHttpBean<T>): R {
        if (respone.isTokenError()) {
            return error(TokenException(respone.getCode(), respone.getMessage()))
        }
        return if (respone.isSuccessful()) {
            handleData(respone.getData())
        } else {
            error(BusinessException(respone.getCode(), respone.getMessage()))
        }
    }

    abstract fun error(ex: Exception): R
    abstract fun handleData(item: T?): R
}